### DISCUSSION TITLE
SPDLOG_LOGGER_CALL with additional data and custom formatting

### DISCUSSION THREAD

**Discussion by AndyDralle** [gabime/spdlog#3012] [Q&A] [answered]
Hi!

So, I have a logger that I want to produce output in JSON format. 

If I call `SPDLOG_LOGGER_CALL(myLogger, myLogLevel, "Some message")`, I suppose I could give my logger a custom formatting string that will take the payload and produce a log message like this: 
`{payload: "Some message"}`

But I would like to introduce some additional data, beyond just the message string. For example, I may have something like this: 
```C++
enum Class Day {
Monday,
Tuesday,
Wednesday,
...
}

struct MessageContext {
Day day;
int count;
}
```

Is there some way for me to include a `MessageContext` object in my logger call, in addition to the payload string? 
Like: 

```C++
MessageContext context{Day::Monday, 5};

SPDLOG_LOGGER_CALL(myLogger, myLogLevel, "Some message", context);
```

And then have my logger produce final output like: 
```C++
{payload: "Some message", day: "Monday", count: 5}
```

I believe this `SPDLOG_LOGGER_CALL` call compiles, but I'm having trouble figuring out what happens to my additional context, once it has become part of the `__VA_ARGS__` in `SPDLOG_LOGGER_CALL`. Is there a way for me to use this additional data to produce my final log output, or is this simply outside the intended use case?

**Comment by tt4g** [comment, marked as answer]
Additional log arguments are formatted by the fmt library, so a custom data formatter for fmt must be defined.
See: https://fmt.dev/9.1.0/api.html#formatting-user-defined-types

However, spdlog does not have an API to output log data as JSON format.
This should be customized by the developer.
See: #1797

**Comment by AndyDralle** [comment]
> Additional log arguments are formatted by the fmt library, so a custom data formatter for fmt must be defined. See: https://fmt.dev/9.1.0/api.html#formatting-user-defined-types

Does this mean that my message would need to include a `{}` placeholder like: 
`SPDLOG_LOGGER_CALL(myLogger, myLogLevel, "Some message {}", context);`?

I suppose fmt does not do anything with additional arguments otherwise.

**Reply by tt4g** [reply]
Yes. `{}` is required.
